File: examples/python/client_collection_utils.py
#!/usr/bin/env python3
"""Collection Utilities Example - Using ekodb_client Python library

This example demonstrates collection utility methods:
- collection_exists: Check if a collection exists
- count_documents: Count documents in a collection
"""

import asyncio
import os
import uuid
from pathlib import Path

from dotenv import load_dotenv

from ekodb_client import Client

# Load environment variables
env_path = Path(__file__).parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")


def is_not_found(error):
    message = str(error).lower()
    return "404" in message or "not found" in message


async def collection_utils_examples():
    """Run Collection Utilities API examples"""
    print("✓ Client created")

    # Create ekoDB client
    client = Client.new(BASE_URL, API_KEY)

    collection = "collection_utils_test_py"
    missing_collection = f"nonexistent_collection_xyz_py_{uuid.uuid4().hex}"
    operation_error = None

    # Start from a deterministic state without touching any collection this
    # example does not own.
    try:
        await client.delete_collection(collection)
    except Exception as exc:
        if not is_not_found(exc):
            raise

    try:
        # Example 1: Check if collection exists (should be false initially)
        print("\n=== Check Collection Exists (Before Creation) ===")
        try:
            exists = await client.collection_exists(collection)
        except Exception as exc:
            print(f"CollectionExists error: {exc}")
            raise
        print(f"Collection '{collection}' exists: {exists}")
        if exists:
            raise AssertionError(f"collection '{collection}' unexpectedly exists")

        # Example 2: Create some test documents
        print("\n=== Creating Test Documents ===")
        for i in range(1, 6):
            record = {
                "name": f"Test Document {i}",
                "index": i,
            }
            try:
                await client.insert(collection, record)
            except Exception as exc:
                print(f"Insert error: {exc}")
                raise
        print("Created 5 test documents")

        # Example 3: Check if collection exists now (should be true)
        print("\n=== Check Collection Exists (After Creation) ===")
        try:
            exists = await client.collection_exists(collection)
        except Exception as exc:
            print(f"CollectionExists error: {exc}")
            raise
        print(f"Collection '{collection}' exists: {exists}")
        if not exists:
            raise AssertionError(f"collection '{collection}' was not created")

        # Example 4: Count documents in collection
        print("\n=== Count Documents ===")
        try:
            count = await client.count_documents(collection)
        except Exception as exc:
            print(f"CountDocuments error: {exc}")
            raise
        print(f"Document count in '{collection}': {count}")
        if count != 5:
            raise AssertionError(f"expected 5 documents, found {count}")

        # Example 5: Check non-existent collection
        print("\n=== Check Non-Existent Collection ===")
        try:
            exists = await client.collection_exists(missing_collection)
        except Exception as exc:
            print(f"CollectionExists error: {exc}")
            raise
        print(f"Collection 'nonexistent_collection_xyz' exists: {exists}")
        if exists:
            raise AssertionError(
                f"collection '{missing_collection}' unexpectedly exists"
            )
    except Exception as exc:  # noqa: BLE001 - preserve the primary example failure
        print(f"Collection utilities error: {exc}")
        operation_error = exc

    # Cleanup: Delete the test collection
    print("\n=== Cleanup ===")
    try:
        try:
            await client.delete_collection(collection)
            print(f"Deleted collection '{collection}'")
        except Exception as exc:
            if not is_not_found(exc):
                raise
        if await client.collection_exists(collection):
            raise AssertionError(
                f"collection '{collection}' still exists after cleanup"
            )
    except Exception as cleanup_error:
        print(f"DeleteCollection error: {cleanup_error}")
        if operation_error is not None:
            operation_error.add_note(f"cleanup also failed: {cleanup_error}")
        else:
            raise

    if operation_error is not None:
        raise operation_error

    print("\n✓ Collection Utilities example complete")


if __name__ == "__main__":
    asyncio.run(collection_utils_examples())

File: examples/typescript/client_collection_utils.ts
/**
 * Collection Utilities Example - Using @ekodb/ekodb-client library
 *
 * This example demonstrates collection utility methods:
 * - collectionExists: Check if a collection exists
 * - countDocuments: Count documents in a collection
 */

import { EkoDBClient } from "@ekodb/ekodb-client";
import * as dotenv from "dotenv";

dotenv.config();

const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";
const API_KEY = process.env.API_BASE_KEY || "a-test-api-key-from-ekodb";

async function main() {
  const client = new EkoDBClient(BASE_URL, API_KEY);
  await client.init();
  console.log("✓ Client created");

  const collection = "collection_utils_test_ts";

  // Example 1: Check if collection exists (should be false initially)
  console.log("\n=== Check Collection Exists (Before Creation) ===");
  try {
    const exists = await client.collectionExists(collection);
    console.log(`Collection '${collection}' exists: ${exists}`);
  } catch (error) {
    console.log(`CollectionExists error: ${error}`);
  }

  // Example 2: Create some test documents
  console.log("\n=== Creating Test Documents ===");
  for (let i = 1; i <= 5; i++) {
    const record = {
      name: `Test Document ${i}`,
      index: i,
    };
    await client.insert(collection, record);
  }
  console.log("Created 5 test documents");

  // Example 3: Check if collection exists now (should be true)
  console.log("\n=== Check Collection Exists (After Creation) ===");
  try {
    const exists = await client.collectionExists(collection);
    console.log(`Collection '${collection}' exists: ${exists}`);
  } catch (error) {
    console.log(`CollectionExists error: ${error}`);
  }

  // Example 4: Count documents in collection
  console.log("\n=== Count Documents ===");
  try {
    const count = await client.countDocuments(collection);
    console.log(`Document count in '${collection}': ${count}`);
  } catch (error) {
    console.log(`CountDocuments error: ${error}`);
  }

  // Example 5: Check non-existent collection
  console.log("\n=== Check Non-Existent Collection ===");
  try {
    const exists = await client.collectionExists("nonexistent_collection_xyz");
    console.log(`Collection 'nonexistent_collection_xyz' exists: ${exists}`);
  } catch (error) {
    console.log(`CollectionExists error: ${error}`);
  }

  // Cleanup: Delete the test collection
  console.log("\n=== Cleanup ===");
  try {
    await client.deleteCollection(collection);
    console.log(`Deleted collection '${collection}'`);
  } catch (error) {
    console.log(`DeleteCollection error: ${error}`);
  }

  console.log("\n✓ Collection Utilities example complete");
}

main().catch((error) => {
  console.error("❌ Error:", error);
  process.exit(1);
});
